Initial implementation of the DNF5 backend

Conan-Kudo opened the pull request:
This pull request is the culmination of a few weeks of _tons_ of trial and error to get an initial backend going for DNF version 5.

It is as much of a straight port from the existing DNF backend as possible, with the hope of future improvements coming afterward.

There might also be leftover logging goop from various iterations... Sorry! 😅 

Fixes: #197 
Fixes: #202 
Fixes: #368 
Fixes: #418 
Fixes: #423 
Fixes: #473 
Fixes: #485 
Fixes: #588 
Fixes: #677 
Fixes: #708 
Fixes: #737 
Fixes: #779 
Fixes: #803 
Fixes: #822

Conan-Kudo commented:
@evan-goode @dmach @jrohel @pkratoch Please take a look, I would appreciate feedback.

Conan-Kudo commented:
I've made a COPR for Fedora Linux users to try out this backend on top of PackageKit 1.3.3.

To try it, just do the following:

1. Enable [the COPR](https://copr.fedorainfracloud.org/coprs/ngompa/PackageKit-DNF5-Preview/): `sudo dnf copr enable ngompa/PackageKit-DNF5-Preview`
2. Upgrade to the new PackageKit build: `sudo dnf upgrade PackageKit\*`

This should install the new build of PackageKit and install the new `PackageKit-backend-dnf5` package.

Conan-Kudo commented:
(FYI, for other distributions who want to backport this once it lands, [here's the patch](https://src.fedoraproject.org/fork/ngompa/rpms/PackageKit/blob/dnf5/f/PackageKit-1.3.3-Initial-DNF5-Backend.patch).)

Conan-Kudo commented:
(Oh, FYI @pwithnall from GNOME Software too)

Conan-Kudo commented:
There is one issue I'm observing at the moment with some testing and I'm trying to troubleshoot: the UpgradeSystem command doesn't seem to correctly load the target distribution release. I'm not sure if this is a 43->rawhide thing or not, but it's annoying that the output isn't somewhat matching what the old backend does.

This is also probably the first real, actual use of the AppStream plugin for libdnf5. I haven't yet tested whether that part works, but if it doesn't, that will need fixing in the dnf5 project.

The rest of the features seem to work as expected.

Conan-Kudo commented:
> Because these commits mention libdnf5 for package and dependency discovery and rpm for actual installation:
> 
> Does that mean this still breaks the `dnf history` functionality, just like the dnf(4) backend does, or are these transactions represented properly within dnf5's view?

No, it's recorded correctly for dnf5's view of the history.

Conan-Kudo commented:
The only thing not fixed right now is the caches issue. The _reason_ it's not fixed is because DNF5 doesn't use versioned caches currently, so there will be collisions and problems if a staged system upgrade in discover/software is overwritten by a manual action in dnf5 in the terminal.

This is fixable, and I'll bring it up in my next meeting with the DNF team so we can come up with an approach that works reasonably here.

evan-goode commented on backends/dnf5/dnf5-backend-thread.cpp:
```diff
@@ -0,0 +1,722 @@
+				pk_backend_job_finished (job);
+				return;
+			} else {
+				g_variant_get (params, "(^as)", &package_ids);
```
This can come out of the `else` right?

gordonmessmer commented:
I've done some interactive testing of this backend. I've only found one thing that looks like an issue, and I'm not sure if it's what @Conan-Kudo is describing as "the caches issue."

With PackageKit/dnf, the gnome-software "Updates" tab will display an icon for an application that provides MetaInfo for which there is a pending update. If I update that application with dnf, gnome-software will refresh the tab, and that application will no longer be listed.

However, with PackageKit/dnf5, the gnome-software "Updates" tab will display the same icon, and if I update that application with dnf, gnome-software will refresh the tab, and that application will still be listed. But if I run "pkgctl refresh", then gnome-software will refresh the tab, and that application will no longer be listed.

I think that suggests that the dnf5 backend is not processing the dnf notification in the same way that the dnf backed does, and the observation that "pkgctl refresh" works as expected suggests that this can be fixed in the backend and does not require a fix in libdnf. However, I'm not sure where that fix belongs yet, so I'll keep looking.

I also ran packagekitd in valgrind and at least initially it looks like the detected leaks are all the same sort of startup allocations seen previously in librpm and libdnf. I'll work on creating a suppressions file and then recheck.

But overall, this looks very good.

Conan-Kudo commented:
> I think that suggests that the dnf5 backend is not processing the dnf notification in the same way that the dnf backed does, and the observation that "pkgctl refresh" works as expected suggests that this can be fixed in the backend and does not require a fix in libdnf. However, I'm not sure where that fix belongs yet, so I'll keep looking.

That sounds like the libdnf5 PK notify plugin might not be working properly (which hasn't changed in this PR at all, code-wise).

Conan-Kudo commented:
Or maybe we're not handling the StateHasChanged refresh trigger thing properly?

aleasto commented:
> Or maybe we're not handling the StateHasChanged refresh trigger thing properly?

That's what I would infer, yes.

Conan-Kudo commented:
Hmm, I need to double check, but I think our refresh function doesn't do anything if we don't pass `force`? GNOME Software and Plasma Discover in Fedora always pass force for refresh, but the notify plugin can't really do that.

Conan-Kudo commented:
Hmm, it _should_ work properly? This is what we're doing in `dnf5_setup_base()`:

```cxx
	auto repo_sack = priv->base->get_repo_sack();
	repo_sack->create_repos_from_system_configuration();
	repo_sack->get_system_repo();

	if (refresh && force) {
		libdnf5::repo::RepoQuery query(*priv->base);
		for (auto repo : query) {
			if (repo->is_enabled()) {
				g_debug("Expiring repository metadata: %s", repo->get_id().c_str());
				repo->expire();
			}
		}
	}

	g_debug("Loading repositories");
	repo_sack->load_repos();
```

gordonmessmer commented:
> That behavior is confusing. Can you reproduce it with Plasma Discover too?

Let's take the external tools out of the picture entirely:

```
$ sudo dnf5 downgrade wireshark\*
$ pkgctl refresh

$ pkgctl list-updates
Starting                                                       [██████████████████████████████████] 100%
▲ wireshark 1:4.6.1-1.fc42.x86_64 [updates]
▲ wireshark-cli 1:4.6.1-1.fc42.x86_64 [updates]
$ sudo dnf5 update wireshark\*
$ pkgctl list-updates
Starting                                                       [██████████████████████████████████] 100%
▲ wireshark 1:4.6.1-1.fc42.x86_64 [updates]
▲ wireshark-cli 1:4.6.1-1.fc42.x86_64 [updates]
$ pkgctl refresh
Refreshing software list                                       [██████████████████████████████████] 100%
✔ Package metadata refreshed
[gordon@vagabond:~/git/PackageKit-dnf5]$ pkgctl list-updates
Starting                                                       [██████████████████████████████████] 100%
```

PackageKit's own client shows wireshark updates pending until after "pkgctl refresh"

mcrha commented:
You can also use RPM-s `rpm-plugin-dbus-announce`, which lets you know when even rpm commands are executed. It can be used like here: https://gitlab.gnome.org/mcrha/gnome-software/-/blob/786bab5f1928920251ee9e93d51953ff9dd9faa4/plugins/dnf5/gs-plugin-dnf5.c#L233

ximion commented:
I will look into it after the 17. when workload is lower and I can give it the attention it deserves. I promise the backend will be in the next PK release. My initial impression is good, but this is a large patch, so I will need some time.

If I do find time prior, you will notice - but otherwise, I'll get to this after defending my PhD.

Conan-Kudo commented:
> I'll get to this after defending my PhD.

Good luck! 🤞🏾

aleasto commented:
The dedicated dnf5 plugin has a critical difference: it notifies PackageKit even if it wasn't running, causing its dbus activation. It was specifically intended to work this way, see discussion at: https://github.com/PackageKit/PackageKit/pull/578#issuecomment-1384781816

I believe all the same reasons still apply, so it should be kept (making rpm-plugin-dbus-announce redundant)

aleasto commented:
> Perhaps instead we need an _rpm_ plugin for PackageKit instead of a _libdnf5_ one? Or something that would otherwise wake up PackageKit from the dbus-announce thing and trigger the refresh.

There could be a tiny daemon listening for rpm-plugin-dbus-announce and waking up PackageKit, yes.

Also, now that I think of it, does the dnf5 plugin now cause a circular loop of `PackageKit transaction -> libdnf -> notify_packagekit plugin -> StateHasChanged -> PackageKit` ?

Conan-Kudo commented:
Oh, yes, it probably does. We currently don't have some details in the transaction database (e.g. no description or comment and uid set to 0) when PackageKit does it.

Example from updating `cppcheck` through PackageKit:

```
Transaction ID : 109
Begin time     : 2026-01-09 14:04:55
Begin rpmdb    : 5de289882dc95fbf03dcf2736a9c785ed42d84d1dfd8243ddda2cf700dba8d89
End time       : 2026-01-09 14:04:56
End rpmdb      : dc72f7ad35fc9e4fff985bd53021d596cec1ff5d5e47e3f7633fb2bb23c62e30
User           : 0 Super User <root>
Status         : Ok
Releasever     : 43
Description    : 
Comment        : 
Packages altered:
  Action   Package                         Reason        Repository
  Upgrade  cppcheck-0:2.19.1-1.fc43.x86_64 External User updates
  Replaced cppcheck-0:2.18.3-1.fc43.x86_64 External User @System
```

vs doing it from the DNF CLI:

```
Transaction ID : 111
Begin time     : 2026-01-09 14:05:58
Begin rpmdb    : 5311b316135c85849301d8e626325d5f475753d9306e7dec54f8a76b2c412e31
End time       : 2026-01-09 14:06:00
End rpmdb      : 17b6ea06be4a947763313bbacb1a22bd0c433244439272a1c3ba896ee846f035
User           : 1000 Neal Gompa <ngompa>
Status         : Ok
Releasever     : 43
Description    : dnf up cppcheck
Comment        : 
Packages altered:
  Action   Package                         Reason        Repository
  Upgrade  cppcheck-0:2.19.1-1.fc43.x86_64 External User updates
  Replaced cppcheck-0:2.18.3-1.fc43.x86_64 External User @System

```

Maybe it might be worth evaluating that so the notify plugin can check and avoid this for obviously PackageKit calls?

dancleaver commented:
Will this make it so KDE Discover respects `dnf versionlock` on Fedora? (related issues listed below)

#677 
#444 
https://bugs.kde.org/show_bug.cgi?id=473607
https://github.com/rpm-software-management/libdnf/issues/1627
https://github.com/rpm-software-management/dnf5/issues/332#issuecomment-2992182792

ximion commented on .github/labeler.yml:
```diff
@@ -28,6 +28,10 @@ library:
 - changed-files:
   - any-glob-to-any-file: backends/dnf/**
 
+'Backend: dnf5 (Fedora Linux)':
```
Then this is alright. Minor version bumps have historically been tied to API additions or major refactoring of PK itself or "whenever we felt like it", not at all to backend changes. So that backend can be dropped at whatever time is most convenient to you and the DNF-ecosystem ;-)
